TESTING.md

```diff
@@ -0,0 +1,79 @@
+# Automated End-to-End Testing
+
+An overview of the End-to-End testing of CHT-Core. 
+
+## Stack overview
+
+### Requirements
+Follow the guide [DEVELOPMENT.md](DEVELOPMENT.MD)
+JDK installed for Selenium.
+Docker to run couchdb.
+
+### Local Run
+
+`grunt e2e` installs and runs chromedriver, starts couchdb in docker, pushes the compiled app to couchdb, starts api, starts sentinel, and then runs protractor tests against your local environment. 
+
+### Github Actions Run
+
+The build process compiles our application. Then installs horticulturalist to run the app. This puts us closer to production. Executes `grunt ci-e2e`. Which then installs and runs chromedriver. Runs the protractor tests against the installed app version. Currently there are 3 jobs that execute in the supported node environments.  
+
+## Protractor Tips
+
+### File Structure
+Tests files should represent a feature within the application. Using `describe` to identify the feature and `it` detail the individual functions of the feature. 
+
+EX: `describe('Users can login )`  `it(with valid credentials)`
+
+### Page Object Model
+We are leveraging the page object model for structure. When identifying elements they should be added to a page object and not within a test file. Add functions that do actions to the page within the page object. Keep expects outside of page objects. The tests should be self documenting. 
+
+### Adding identifiers
+In some cases adding a unique identifier to an element may be necessary. This could be a piece of data related to the element or a unique name. To do this the app code can be modified to add a `test-` attribute.  
+
+Ex:  `attr.test-id="{{ msg.key }}" ` Will add a `test-id` attribute with the data from the app. 
+
+Then it can be consumed in the test by getting an element by css. EX: ``element(by.css(`#message-list li[test-id="${identifier}"]`)),``
+
+
```

Comment:
```suggestion
Adding a test identifier a good option for cases where a CSS selector would otherwise be fragile such as selecting based on an assumed element structure or selecting on CSS classes intended for visual design that may change.
```